src/steamship/agents/service/agent_service.py

```diff
@@ -146,6 +148,20 @@ def run_agent(self, agent: Agent, context: AgentContext):
 
         while not isinstance(action, FinishAction):
             self.run_action(agent=agent, action=action, context=context)
+
+            # If this action wishes to end the loop, comply with that wish
+            if action.return_direct:
+                logging.info(
```

Comment:
curious: do you think we need this log statement for full coverage?  WDYT about just adding "is_final" or similar to the existing "next tool" log?  that might lead to more readable code (no floating `break`) and fewer lines of logging code?